Validate IPAM config before handing it over to libnetwork

akerouanton opened the pull request:
**- What I did**

Prior to this change, IPAM config was never validated by the API. Some checks are done by the CLI, but it's not exhaustive (and doesn't cover API users anyway). Some of these errors might be caught early by libnetwork (eg. when the network is created), and other errors would surface only when connecting a container to a network with faulty IPAM. In both cases, the API would return a 500 instead of returning a proper 400 at creation time.

Although the `NetworkCreate` endpoint might already return warnings, these are never displayed by the CLI. As such, it was decided during a maintainer's call to return validation errors _for all API versions_.

Prior to #44968, the daemon would happily accept an `--ip-range` with a larger mask than its parent subnet. In such case, the IPAM allocator would implicitly ignore the ip-range (not because it detected it was a wrong value, but due to the way computation was done). As such, the 2nd commit explicitly discard the ip-range and replace it with the subnet value when it detects it's a bad value. This error would be caught by validation added in the first commit, so it's only useful for network restored from state stored on-disk.

One such example can be seen in docker/for-mac#6870. The user creates a network with an ip-range bigger than its subnet, and with some host bits set. When they try to create a container connected to that network, the daemon returns this error: `docker: Error response from daemon: invalid bit range [0, 16777215).`. This defensive check is right, but the user should be warned of bad IPAM config earlier and in a more graceful way.

**- How to verify it**

```
$ docker network create --subnet 172.92.0.0/16 --ip-range 172.92.0.0/8 --gateway 172.92.0.1 main2
Error response from daemon: invalid network config:
* invalid ip-range 172.92.0.0/8: CIDR block is bigger than its parent subnet 172.92.0.0/16
* invalid ip-range 172.92.0.0/8: it should be 172.0.0.0/8
```

**- Description for the changelog**

* Validate IPAM config before creating a network.

**- A picture of a cute animal (not mandatory but encouraged)**

<img src="https://encrypted-tbn0.gstatic.com/licensed-image?q=tbn:ANd9GcTwhRXEoI2UZS4kGW3diALP-9vVT0np31Db8Z5WsT-1yyWqx3ovqROQzneHzzmo12i3_1ijDpHBAwwCVEg"  width="50%" height="50%">

akerouanton committed:
api: Validate IPAM config before creating a network

Currently, IPAM config is never validated by the API. Some checks
are done by the CLI, but they're not exhaustive. And some of these
misconfigurations might be caught early by libnetwork (ie. when the
network is created), and others only surface when connecting a container
to a misconfigured network. In both cases, the API would return a 500.

Although the `NetworkCreate` endpoint might already return warnings,
these are never displayed by the CLI. As such, it was decided during a
maintainer's call to return validation errors _for all API versions_.

Signed-off-by: Albin Kerouanton <albinker@gmail.com>
Signed-off-by: Sebastiaan van Stijn <github@gone.nl>

akerouanton committed:
ipam: Replace ChildSubnet with parent Subnet when its mask is bigger

Prior to moby/moby#44968, libnetwork would happily accept a ChildSubnet
with a bigger mask than its parent subnet. In such case, it was
producing IP addresses based on the parent subnet, and the child subnet
was not allocated from the address pool.

This commit automatically fixes invalid ChildSubnet for networks stored
in libnetwork's datastore.

Signed-off-by: Albin Kerouanton <albinker@gmail.com>

neersighted commented:
Discussed in the maintainers call / @akerouanton is making the below changes:
* Semantics for existing networks look good
* We want to hard fail in new major versions of the engine, and soft fail in older versions:
  * This is in place of the current behavioral difference based on API version
  * This should be accomplished by a two-commit dance: one commit introduces the warning, another changes it to an error.
  * Cherry-picks to older branches should exclude the commit that makes this an error.

neersighted reviewed (changes requested):
Discussed in the maintainers call / @akerouanton is making the below changes:

* Semantics for existing networks look good
* We want to hard fail in new major versions of the engine, and soft fail in older versions:
    * This is in place of the current behavioral difference based on API version.
    * This should be accomplished by a two-commit dance: one commit introduces the warning, another changes it to an error.
    * Cherry-picks to older branches should exclude the commit that makes this an error.

thaJeztah commented on api/server/router/network/network_routes.go:
```diff
@@ -216,16 +219,22 @@ func (n *networkRouter) postNetworkCreate(ctx context.Context, w http.ResponseWr
+		for _, err := range errs.WrappedErrors() {
+			warnings = append(warnings, err.Error())
+		}
+	}
```
I think I'd rather not break the "error" abstraction here, and instead make `` return a regular error (`return `). go-multierror already has a default format to combine multiple errors, which would look something like this;

```
    2 errors occurred:
        	* subnet blablabla has some bits set in its host fragment
        	* IPv6 has not been enabled for this network, but an IPv6 subnet has been provided
```

But if we really want to define our own formatting, we can set the `ErrorFormat` instead, and implement our own; https://github.com/moby/moby/blob/97455cc31ffa08078db6591f018256ed59c35bbc/vendor/github.com/hashicorp/go-multierror/multierror.go#L15-L22

The default one looks like this; https://github.com/moby/moby/blob/97455cc31ffa08078db6591f018256ed59c35bbc/vendor/github.com/hashicorp/go-multierror/format.go#L12-L27

And make this code;

```go
	if err := validateIPAM(create.IPAM, create.EnableIPv6); err != nil {
		warnings = append(warnings, err.Error())
	}
```

thaJeztah commented on api/server/router/network/network_routes.go:
```diff
@@ -236,14 +245,133 @@ func (n *networkRouter) postNetworkCreate(ctx context.Context, w http.ResponseWr
 	return httputils.WriteJSON(w, http.StatusCreated, nw)
 }
 
+type ipFamily int
+
+const (
+	ip4 ipFamily = 4
+	ip6 ipFamily = 6
+)
+
+func validateIPAM(ipam *network.IPAM, enableIPv6 bool) *multierror.Error {
```
I'm a bit on the fence on having all this validation code in the API router. Given that we're already importing libnetwork, perhaps it'd make sense to move the validation code there (not sure where the right place would be inside libnetwork, perhaps in the `ipam`  or `ipamapi` package.

We can either have the function exported (and call it here; `ipam.Validate(...)`), or perhaps as part of the final commit (which turns it into an error), move it into `backend.CreateNetwork()`. That way calling that function would also validate before creating (regardless of where it's called from).

akerouanton replied:
> I'm a bit on the fence on having all this validation code in the API router.

I think the best would be to add IPAM validation call to `(*Daemon).CreateNetwork()`. However, the following method from swarmkit would prevent converting errors into warnings when the network is created from a non-leader manager node, as it returns either the network created or an error, but not both: 

https://github.com/moby/swarmkit/blob/ad0f3ae162fa30d0710105211d37996d5ec38b27/api/control.pb.go#L6678-L6685

Also, there's already some sort of network validation in the daemon pkg (ie. `d.containerStart() -> d.initializeNetworking() -> d.allocateNetwork() -> d.connectToNetwork() -> d.updateNetworkConfig() -> validateNetworkingConfig()`).

> Given that we're already importing libnetwork, perhaps it'd make sense to move the validation code there

Currently the API only uses constants and errors from libnetwork packages. If we call IPAM validation helper from the API due to my previous point, these helpers would need to either:

1. Rely on the API type `IPAMConfig`, and thus import it into libnetwork -- that'd be a bad separation of concern IMO ;
2. Convert API type `IPAMConfig` into libnetwork `IpamConf`. And then, errors would have to either reference fields from the latter struct and re-translated to match the former -- that'd make the validation layer more complex with no real gains. Or modeled after `IPAMConfig`, but then we're back at bad separation of concerns ;

Also, I think we should make sure as much as possible that libnetwork data structures never end up in an invalid state -- that's the crux of the original issue I'm solving here. Maintaining libnetwork is way harder if we can't trust this rule is enforced. This also discards option 2.

thaJeztah replied:
> Currently the API only uses constants and errors from libnetwork packages. If we call IPAM validation helper from the API due to my previous point, these helpers would need to either:
>
> - Rely on the API type IPAMConfig, and thus import it into libnetwork -- that'd be a bad separation of concern IMO ;
> - Convert API type IPAMConfig into libnetwork IpamConf. And then, errors would have to either reference fields from the latter struct and re-translated to match the former -- that'd make the validation layer more complex with no real gains. Or modeled after IPAMConfig, but then we're back at bad separation of concerns ;
> - Also, I think we should make sure as much as possible that libnetwork data structures never end up in an invalid state -- that's the crux of the original issue I'm solving here. Maintaining libnetwork is way harder if we can't trust this rule is enforced. This also discards option 2.

- Looking at the validation we have in this PR, we only need `create.IPAM`, and `create.EnableIPv6`.
- `backend.CreateNetwork()` is implemented by `daemon.CreateNetwork()` / `daemon.createNetwork()`
- .. which already imports `"github.com/docker/docker/libnetwork/ipamapi"`
- .. and has a `getIpamConfig()` function (which can produce an error for invalid configuration)

And on the calling site (`network_routes.go`), we already have some checking for error-types returned by the backend (e.g. `libnetwork.NetworkNameError`) to turn errors into warnings.

So my thinking is to

- have the validation either in the backend (`daemon.createNetwork()` -> `getIpamConfig()`) itself (or in the `ipamapi` package, and called from there)
- make it return a sentinel error (for example, `libnetwork.InvalidIPAMConfig`)
- ^^ Ideally that error would implement `errdefs.InvalidParam`

For the current release, we catch that (`libnetwork.InvalidIPAMConfig`), and turn them into a warning, and for the next release, we remove that check, and return the error as-is (producing a 4XX status if it implements `errdefs.InvalidParam`)


Would that work?

thaJeztah commented on api/server/router/network/network_routes.go:
```diff
@@ -236,14 +245,133 @@ func (n *networkRouter) postNetworkCreate(ctx context.Context, w http.ResponseWr
 	return httputils.WriteJSON(w, http.StatusCreated, nw)
 }
 
+type ipFamily int
+
+const (
+	ip4 ipFamily = 4
+	ip6 ipFamily = 6
+)
```
Looks like we're always assigning these values internally, so we may as well save the pain of having to format this into a human-readable string;

```go
type ipFamily string

const (
	ip4 ipFamily = "IPv4"
	ip6 ipFamily = "IPv6"
)
```

thaJeztah commented on api/server/router/network/network_routes_test.go:
```diff
@@ -0,0 +1,121 @@
+					AuxAddress: map[string]string{"DefaultGatewayIPv4": "dummy"},
+				}},
+			},
+			expectedErrors: []string{
+				"ip-range \"foobar\" is an invalid prefix",
+				"gateway \"barbaz\" is an invalid address",
+				"auxiliary address \"dummy\" is an invalid address",
+			},
```
Can use string-literals (backticks) for these; that way there's no need to escape the quotes;

```go
expectedErrors: []string{
	`ip-range "foobar" is an invalid prefix`,
	`gateway "barbaz" is an invalid address`,
	`auxiliary address "dummy" is an invalid address`,
},
```

thaJeztah commented on api/server/router/network/network_routes_test.go:
```diff
@@ -0,0 +1,121 @@
+
+			errs := validateIPAM(&tc.ipam, tc.ipv6)
+			for _, err := range tc.expectedErrors {
+				assert.ErrorContains(t, errs, err)
```
Use `assert.Check()` here so that we don't fail the test on the first missed error, but continue with the other ones;


```go
assert.Check(t, is.ErrorContains(errs, err))
```

thaJeztah reviewed (changes requested):
Thanks; I left some suggestions; let me know what you think.

I committed some of my suggestions in a branch, and just opened a PR against your fork in case that helps you seeing them as a whole (except for moving the validation to libnetwork if you agree on that);

- https://github.com/akerouanton/docker/pull/2

corhere commented on api/server/router/network/network_routes.go:
```diff
@@ -236,14 +245,133 @@ func (n *networkRouter) postNetworkCreate(ctx context.Context, w http.ResponseWr
+	for _, cfg := range ipam.Config {
+		subnet, err := netip.ParsePrefix(cfg.Subnet)
+		if err != nil {
+			errs = multierror.Append(errs, fmt.Errorf("subnet %q is an invalid prefix", cfg.Subnet))
```
Preserve stack for what? I'm not seeing how the ability to pinpoint the code location where user input failed a particular validation check will be of much use. I could _maybe_ see some use for wrapping the topmost multierror to preserve a stack trace down to the `validateIPAM` function, but not the inner errors.

thaJeztah replied:
Yeah, not critical (mostly because we were already using pkg/errors for various other errors here, and thought it probably wouldn't _hurt_)

thaJeztah commented on api/server/router/network/network_routes.go:
```diff
@@ -236,14 +243,131 @@ func (n *networkRouter) postNetworkCreate(ctx context.Context, w http.ResponseWr
+	ip6 ipFamily = 6
+)
+
+func validateIPAM(ipam *network.IPAM, enableIPv6 bool) *multierror.Error {
```
Left similar suggestions; I opened a PR against @akerouanton's fork with my (draft) suggestions; https://github.com/akerouanton/docker/pull/2 (feel free to review those changes if you want to 😅)

akerouanton commented:
I did a few things in the last force push:

* The data validation logic is now moved to the `daemon` package ;
* Improved error messages to make them less chatty while retaining important info ;
* Rebased on top of #46099, to see how the formatted `message` field is shown on the CLI:

```
$ docker network create --subnet 172.92.0.0/16 --ip-range 172.92.0.0/8 --gateway 172.92.0.1 main2
Error response from daemon: 2 errors occurred:
	* invalid ip-range 172.92.0.0/8: CIDR block is bigger than its parent subnet 172.92.0.0/16
	* invalid ip-range 172.92.0.0/8: some bits are set in the host fragment
```

* And finally, `github.com/pkg/errors` is replaced by stdlib's `errors` in `daemon/network.go`, to make `errors.Join()` available

akerouanton committed:
api/t/network: Move IPAM types to their own file

Signed-off-by: Albin Kerouanton <albinker@gmail.com>

thaJeztah reviewed (approved):
LGTM if CI goes green 👍
